### DISCUSSION TITLE
How to get superset datasource permissions through api.

### DISCUSSION THREAD

**Discussion by sundar68** [apache/superset#28488] [Q&A / Help]
How to get superset datasource permissions through api. 

Unable to get datasource permissions id's to assign to newly created role through superset api.

Created new role using this api:
```
POST
body:{
 "name":"NEW_ROLE"
}
http://localhost:8088/api/v1/security/roles/
```

Trying to provide datasource permissions to the role using api:
(How do I get the datasource permission id's)
```
POST
body:
{
  "permission_view_menu_ids": [
    0
  ]
}
http://localhost:8088/api/v1/security/roles/{role_id}/permissions
``` 

In this API datasource permissions are not coming.
```
http://localhost:8088/api/v1/security/permissions/
```

**Comment by GaboBazeth** [comment]
Hello, im having the same issue as you, i'm not beeing able to find the permissions id's through any method in the api.

were you able to find a solution?

**Comment by petermorgan** [comment]
Hi Sundar, I have solved it for my problem and I hope it helps you.

I found the:

`GET /api/v1/security/permissions-resources/`

api is able to return the ids of the permissions, for example:
```

   {
      "id": 326,
      "permission": {
        "name": "datasource_access"
      },
      "view_menu": {
        "name": "[Elevate Insight].[Contracts](id:118)"
      }
    },
```
Then you can:

`POST /api/v1/security/roles/{role_id}/permissions   `

passing in the 326 example above and it works.

I could not get the filtering to work on the:

`GET /api/v1/security/permissions-resources/`

operation so I had to fetch all of them and filter client side.

**Reply by Pratyush1403** [reply]
@petermorgan i am just getting the following response for this route:

GET /api/v1/security/permissions-resources/

> {
>     "count": 437,
>     "description_columns": {},
>     "ids": [
>         1,
>         2,
>         3,
>         4,
>         5,
>         6,
>         7,
>         8,
>         9,
>         10,
>         11,
>         12,
>         13,
>         14,
>         15,
>         16,
>         17,
>         18,
>         19,
>         22
>     ],
>     "label_columns": {
>         "id": "Id",
>         "permission.name": "Permission Name",
>         "view_menu.name": "View Menu Name"
>     },
>     "list_columns": [
>         "id",
>         "permission.name",
>         "view_menu.name"
>     ],
>     "list_title": "List Permission View",
>     "order_columns": [
>         "id",
>         "permission.name",
>         "view_menu.name"
>     ],
>     "result": [
>         {
>             "id": 1,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "SavedQuery"
>             }
>         },
>         {
>             "id": 2,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "SavedQuery"
>             }
>         },
>         {
>             "id": 3,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "CssTemplate"
>             }
>         },
>         {
>             "id": 4,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "CssTemplate"
>             }
>         },
>         {
>             "id": 5,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "ReportSchedule"
>             }
>         },
>         {
>             "id": 6,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "ReportSchedule"
>             }
>         },
>         {
>             "id": 7,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Chart"
>             }
>         },
>         {
>             "id": 8,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "Chart"
>             }
>         },
>         {
>             "id": 9,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Annotation"
>             }
>         },
>         {
>             "id": 10,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "Annotation"
>             }
>         },
>         {
>             "id": 11,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Dataset"
>             }
>         },
>         {
>             "id": 12,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "Dataset"
>             }
>         },
>         {
>             "id": 13,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Log"
>             }
>         },
>         {
>             "id": 14,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "Log"
>             }
>         },
>         {
>             "id": 15,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Dashboard"
>             }
>         },
>         {
>             "id": 16,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "Dashboard"
>             }
>         },
>         {
>             "id": 17,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Database"
>             }
>         },
>         {
>             "id": 18,
>             "permission": {
>                 "name": "can_write"
>             },
>             "view_menu": {
>                 "name": "Database"
>             }
>         },
>         {
>             "id": 19,
>             "permission": {
>                 "name": "can_read"
>             },
>             "view_menu": {
>                 "name": "Query"
>             }
>         },
>         {
>             "id": 22,
>             "permission": {
>                 "name": "can_csv_upload"
>             },
>             "view_menu": {
>                 "name": "Database"
>             }
>         }
>     ]
> }

i am not getting any permission regarding datasource access. am i doing anything wrong?

**Reply by peter-morgan-imperatives** [reply]
Hi @Pratyush1403 I suspect you may need to page through the results. That is, do:

GET /api/v1/security/permissions-resources/?q=(page:0,page_size:100)

Then do:

GET /api/v1/security/permissions-resources/?q=(page:1,page_size:100)

etc..

repeatedly until you have all the permissions.

You then get things like this:

..
{
  "id" : 923,
  "permission" : {
    "name" : "datasource_access"
  },
  "view_menu" : {
    "name" : "[Elevate Insight].[Rated bolton](id:677)"
  }
}, {
  "id" : 924,
  "permission" : {
    "name" : "datasource_access"
  },
  "view_menu" : {
    "name" : "[Elevate Insight].[Contracts](id:678)"
  }
},
...

In my case they are quite high ids so on page 9.

I hope that helps.

Kind regards 

Pete.

**Reply by Pratyush1403** [reply]
Hey @petermorgan 
Thanks a lot. This solved my problem.

Regards

Pratyush
